### DISCUSSION TITLE
FastAPI is not using enough threads

### DISCUSSION THREAD

**Discussion by nilansaha** [fastapi/fastapi#8587] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
..
```

### Description

When I send too many requests to my fastapi application it spends most of the time doing nothing which I think is because of threading switching and there being not enough threads in the pool. However, the CPU usage is nowhere near 100% and the memory is less than 10% I am wondering how can I force fastapi to use more threads ? I could not find any resource to set the thread to a higher number in the latest version and what is the default number of threads ?

### Operating System

Linux

### FastAPI Version

0.73

### Python Version

3.9

**Comment by odiseo0** [comment]
Maybe you should add more workers?

**Comment by nilansaha** [comment]
> Maybe you should add more workers?

I mean thats one approach but I want to use more threads. Is there a way to specify the max threads ?

**Comment by raphaelauv** [comment]
How do you run fastapi, with uvicorn , gunicorn ?

**Comment by ericman93** [comment]
I too have this issue
I'm running fastapi with gunicron with only 1 worker, and looking at the profiler I see that most of time is in the await part when I run multiple requests

**Comment by Kludex** [comment]
You need to retrieve the default `CapacityLimiter` using [`current_default_thread_limiter()`](https://anyio.readthedocs.io/en/stable/api.html?highlight=default%20capaciter#anyio.to_thread.current_default_thread_limiter) and change the number of `total_tokens` (attribute).

Default is 40.

**Comment by nilansaha** [comment]
> You need to retrieve the default `CapacityLimiter` using [`current_default_thread_limiter()`](https://anyio.readthedocs.io/en/stable/api.html?highlight=default%20capaciter#anyio.to_thread.current_default_thread_limiter) and change the number of `total_tokens` (attribute).
> 
> Default is 40.

I believe I am using an older version that doesn't use AnyIO but can I ask where you found that the default is 40 ? Would be great if you can paste a link or someting

**Comment by Kludex** [comment]
You mentioned that you're using 0.73...

Just do what I said above but instead of changing the value of `total_tokens`, just print it. You'll see the default value.

**Comment by Kludex** [comment, marked as answer]
Just in case...

```python
import anyio
from fastapi import FastAPI


app = FastAPI()


@app.on_event("startup")
async def startup():
    limiter = anyio.to_thread.current_default_thread_limiter()
    limiter.total_tokens = 100


@app.get("/")
async def home():
    print(anyio.to_thread.current_default_thread_limiter().total_tokens)  # you'll see 100
```

**Comment by sid203** [comment]
I don't see any improvement in the CPU usage after changing `total_tokens` to 100. I am guessing this might be because my task is CPU bound and in python threads are useful for IO bound tasks?
